// @atlaskit/eslint-plugin-design-system/dist/es2019/rules/no-separator-with-list-elements/index.js
import { isNodeOfType } from 'eslint-codemod-utils';
import { createLintRule } from '../utils/create-rule';
const separatorAsCombinationNotAllowed = 'The combination of `separator` with `as="li"`, `as="ol"`, or `as="dl"` is not allowed.';
const rule = createLintRule({
  meta: {
    name: 'no-separator-with-list-elements',
    type: 'suggestion',
    docs: {
      description: 'Warn when the `separator` prop is used with `as="li"`, `as="ol"`, or `as="dl"` in the Inline component.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      separatorAsCombinationNotAllowed
    }
  },
  create(context) {
    const inlineComponentNames = [];
    return {
      ImportDeclaration(node) {
        if (node.type === 'ImportDeclaration' && (node.source.value === '../src' || node.source.value === '@atlaskit/primitives')) {
          node.specifiers.forEach(specifier => {
            if (specifier.type === 'ImportSpecifier' && 'name' in specifier.imported && specifier.imported.name === 'Inline') {
              inlineComponentNames.push(specifier.local.name);
            }
          });
        }
      },
      JSXElement(node) {
        if (!isNodeOfType(node, 'JSXElement') || !isNodeOfType(node.openingElement.name, 'JSXIdentifier')) {
          return;
        }
        const componentName = node.openingElement.name.name;
        if (!inlineComponentNames.includes(componentName)) {
          return;
        }
        const inlineProps = node.openingElement.attributes.filter(attr => isNodeOfType(attr, 'JSXAttribute') && isNodeOfType(attr.name, 'JSXIdentifier'));
        const separatorProp = inlineProps.find(attr => attr.name.name === 'separator');
        const asProp = inlineProps.find(attr => attr.name.name === 'as');
        if (separatorProp && asProp && asProp.value && isNodeOfType(asProp.value, 'Literal') && ['li', 'ol', 'dl'].includes(asProp.value.value)) {
          context.report({
            node: node,
            messageId: 'separatorAsCombinationNotAllowed'
          });
        }
      }
    };
  }
});
export default rule;
