// @atlaskit/eslint-plugin-design-system/dist/esm/rules/no-html-textarea/index.js
import { createLintRule } from '../utils/create-rule';
import { JSXElement as _JSXElement, StyledComponent } from './node-types';
var rule = createLintRule({
  meta: {
    name: 'no-html-textarea',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML textarea elements in favor of the Atlassian Design System textarea component.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlTextarea: "This <{{ name }}> should be replaced with a textarea component from the Atlassian Design System. ADS components include event tracking, ensure accessible implementations, and provide access to ADS styling features like design tokens."
    }
  },
  create: function create(context) {
    return {
      // transforms styled.<textarea>(...) usages
      CallExpression: function CallExpression(node) {
        StyledComponent.lint(node, {
          context: context
        });
      },
      // transforms <textarea css={...}> usages
      JSXElement: function JSXElement(node) {
        _JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
export default rule;
