// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/use-primitives-text/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createRule = require("../utils/create-rule");
var _errorBoundary = require("../utils/error-boundary");
var _config = require("./config");
var _transformers = require("./transformers");
var textDocsUrl = 'https://atlassian.design/components/primitives/text';
var rule = (0, _createRule.createLintRule)({
  meta: {
    name: 'use-primitives-text',
    type: 'suggestion',
    fixable: 'code',
    hasSuggestions: true,
    schema: [{
      type: 'object',
      properties: {
        failSilently: {
          type: 'boolean'
        },
        inheritColor: {
          type: 'boolean'
        },
        enableUnsafeAutofix: {
          type: 'boolean'
        },
        enableUnsafeReport: {
          type: 'boolean'
        },
        patterns: {
          maxLength: _config.PATTERNS.length,
          type: 'array',
          items: {
            type: 'string',
            enum: _config.PATTERNS
          },
          uniqueItems: true
        }
      },
      additionalProperties: false
    }],
    docs: {
      description: 'Encourage the usage of text components.',
      recommended: false,
      severity: 'warn'
    },
    messages: {
      preferPrimitivesText: "This element can be replaced with a \"Text\" primitive. See ".concat(textDocsUrl, " for additional guidance."),
      preferPrimitivesStackedText: "These paragraphs can be replaced with a \"Text\" and \"Stack\" primitives. See ".concat(textDocsUrl, " for additional guidance.")
    }
  },
  create: function create(context) {
    // TODO: JFP-2823 - this type cast was added due to Jira's ESLint v9 migration
    var config = (0, _config.getConfig)(context.options[0]);
    return (0, _errorBoundary.errorBoundary)({
      'JSXElement[openingElement.name.name=span]': function JSXElementOpeningElementNameNameSpan(node) {
        return _transformers.SpanElements.lint(node, {
          context: context,
          config: config
        });
      },
      'JSXElement[openingElement.name.name=p]': function JSXElementOpeningElementNameNameP(node) {
        return _transformers.ParagraphElements.lint(node, {
          context: context,
          config: config
        });
      },
      'JSXElement[openingElement.name.name=strong]': function JSXElementOpeningElementNameNameStrong(node) {
        return _transformers.StrongElements.lint(node, {
          context: context,
          config: config
        });
      },
      'JSXElement[openingElement.name.name=em]': function JSXElementOpeningElementNameNameEm(node) {
        return _transformers.EmphasisElements.lint(node, {
          context: context,
          config: config
        });
      }
    }, config);
  }
});
var _default = exports.default = rule;
