// eslint-codemod-utils/dist/cjs/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertJSXAttribute = exports.closestOfType = exports.getNodeAfterComment = exports.getFirstCommentInFile = exports.insertAtStartOfFile = exports.getIdentifierInParentScope = exports.isNodeOfType = exports.node = void 0;
var node_1 = require("./node");
Object.defineProperty(exports, "node", { enumerable: true, get: function () { return node_1.node; } });
var is_node_of_type_1 = require("./is-node-of-type");
Object.defineProperty(exports, "isNodeOfType", { enumerable: true, get: function () { return is_node_of_type_1.isNodeOfType; } });
var get_identifier_in_parent_scope_1 = require("./get-identifier-in-parent-scope");
Object.defineProperty(exports, "getIdentifierInParentScope", { enumerable: true, get: function () { return get_identifier_in_parent_scope_1.getIdentifierInParentScope; } });
var insert_at_start_of_file_1 = require("./insert-at-start-of-file");
Object.defineProperty(exports, "insertAtStartOfFile", { enumerable: true, get: function () { return insert_at_start_of_file_1.insertAtStartOfFile; } });
var get_first_comment_in_file_1 = require("./get-first-comment-in-file");
Object.defineProperty(exports, "getFirstCommentInFile", { enumerable: true, get: function () { return get_first_comment_in_file_1.getFirstCommentInFile; } });
var get_node_after_comment_1 = require("./get-node-after-comment");
Object.defineProperty(exports, "getNodeAfterComment", { enumerable: true, get: function () { return get_node_after_comment_1.getNodeAfterComment; } });
var closest_of_type_1 = require("./closest-of-type");
Object.defineProperty(exports, "closestOfType", { enumerable: true, get: function () { return closest_of_type_1.closestOfType; } });
var insert_jsx_attribute_1 = require("./insert-jsx-attribute");
Object.defineProperty(exports, "insertJSXAttribute", { enumerable: true, get: function () { return insert_jsx_attribute_1.insertJSXAttribute; } });
__exportStar(require("./utils"), exports);
