// @atlaskit/eslint-plugin-design-system/dist/es2019/rules/no-emotion-primitives/index.js
import { literal } from 'eslint-codemod-utils';
import { createLintRule } from '../utils/create-rule';
const schema = {
  type: 'array',
  items: [{
    type: 'object',
    properties: {
      autofix: {
        type: 'boolean'
      }
    },
    additionalProperties: false
  }]
};
const defaultConfig = {
  autofix: false
};
function readConfig(context) {
  const [config] = context.options;
  return Object.assign({}, defaultConfig, config);
}
export const rule = createLintRule({
  meta: {
    name: 'no-emotion-primitives',
    type: 'problem',
    fixable: 'code',
    docs: {
      description: 'Ensures usage of Compiled Primitives import instead of Emotion entrypoint.',
      severity: 'warn',
      recommended: false
    },
    messages: {
      'no-emotion-primitives': 'Use @atlaskit/primitives/compiled instead of @atlaskit/primitives. Refer to go/akcss for more information.'
    },
    schema
  },
  create(context) {
    const config = readConfig(context);
    return {
      ImportDeclaration(node) {
        const importSource = node.source.value;
        if (importSource !== '@atlaskit/primitives') {
          return;
        }
        context.report({
          node: node.source,
          messageId: 'no-emotion-primitives',
          fix(fixer) {
            if (!config.autofix) {
              return null;
            }
            const newSource = literal('@atlaskit/primitives/compiled');
            return fixer.replaceText(node.source, newSource.raw || "'@atlaskit/primitives/compiled'");
          }
        });
      }
    };
  }
});
export default rule;
