// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/no-html-range/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createRule = require("../utils/create-rule");
var _nodeTypes = require("./node-types");
var rule = (0, _createRule.createLintRule)({
  meta: {
    name: 'no-html-range',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML range elements in favor of the Atlassian Design System range component.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlRange: "This <{{ name }}> should be replaced with a range component from the Atlassian Design System. ADS components include event tracking, ensure accessible implementations, and provide access to ADS styling features like design tokens."
    }
  },
  create: function create(context) {
    return {
      // transforms <input type="range" css={...}> usages
      JSXElement: function JSXElement(node) {
        _nodeTypes.JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
var _default = exports.default = rule;
