// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/use-tokens-typography/config/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ruleSchema = exports.getConfig = void 0;
/* eslint-disable @atlassian/tangerine/import/entry-points */

var ruleSchema = exports.ruleSchema = {
  type: 'array',
  items: {
    type: 'object',
    properties: {
      failSilently: {
        type: 'boolean'
      },
      shouldEnforceFallbacks: {
        type: 'boolean'
      },
      enableUnsafeAutofix: {
        type: 'boolean'
      },
      patterns: {
        type: 'array',
        items: {
          type: 'string',
          enum: ['style-object', 'font-weight', 'font-family', 'untokenized-properties', 'banned-properties', 'restricted-capitalisation']
        }
      }
    }
  }
};
var defaultConfig = {
  failSilently: false,
  shouldEnforceFallbacks: false,
  enableUnsafeAutofix: false,
  patterns: ['style-object', 'font-weight', 'font-family', 'untokenized-properties', 'banned-properties', 'restricted-capitalisation']
};
var getConfig = exports.getConfig = function getConfig(overrides) {
  return Object.assign({}, defaultConfig, overrides);
};
