// @atlaskit/eslint-plugin-design-system/dist/esm/rules/use-heading-level-in-spotlight-card/index.js
import { isNodeOfType } from 'eslint-codemod-utils';
import { getScope } from '@atlaskit/eslint-utils/context-compat';
import { createLintRule } from '../utils/create-rule';
import { getImportName } from '../utils/get-import-name';
export var headingLevelRequiredSuggestionText = 'Add a `headingLevel` that is of a contextually relevant level.';
var rule = createLintRule({
  meta: {
    name: 'use-heading-level-in-spotlight-card',
    type: 'suggestion',
    fixable: 'code',
    docs: {
      description: 'Inform developers of eventual requirement of `headingLevel` prop in `SpotlightCard` component. The heading level should be the appropriate level according to the surrounding context.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      headingLevelRequired: headingLevelRequiredSuggestionText
    }
  },
  create: function create(context) {
    return {
      JSXElement: function JSXElement(node) {
        if (!isNodeOfType(node, 'JSXElement')) {
          return;
        }
        if (!isNodeOfType(node.openingElement.name, 'JSXIdentifier')) {
          return;
        }

        // Get the name of the SpotlightCard import
        var spotlightCardImportName = getImportName(getScope(context, node), '@atlaskit/onboarding', 'SpotlightCard');
        if (node.openingElement.name.name === spotlightCardImportName) {
          // and if `heading` exists and `headingLevel` prop does not exist
          var spotlightCardProps = node.openingElement.attributes.filter(function (attr) {
            return isNodeOfType(attr, 'JSXAttribute');
          }).filter(function (attr) {
            return attr.name.type === 'JSXIdentifier';
          });
          var heading = spotlightCardProps.find(function (attr) {
            return attr.name.name === 'heading';
          });
          var headingLevel = spotlightCardProps.find(function (attr) {
            return attr.name.name === 'headingLevel';
          });
          if (heading && !headingLevel) {
            context.report({
              node: node,
              messageId: 'headingLevelRequired'
            });
          }
        }
      }
    };
  }
});
export default rule;
