// @atlaskit/form/dist/es2019/range-field.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import Field from './field';
/**
 * __Range field__
 *
 * A range field is where a user can submit a range input as a part of a form.
 *
 * - [Examples](https://atlaskit.atlassian.com/packages/design-system/form/docs/fields#rangefield-reference)
 * - [Code](https://atlaskit.atlassian.com/packages/design-system/form/docs/fields#rangefield-reference)
 * - [Usage](https://atlaskit.atlassian.com/packages/design-system/form/docs/fields#rangefield-reference)
 */
const RangeField = ({
  children,
  defaultValue,
  id,
  isDisabled,
  label,
  name,
  ...strippedProps
}) => /*#__PURE__*/React.createElement(Field, _extends({}, fg('platform_design-system-team_range-field-spread') ? {} : strippedProps, {
  defaultValue: defaultValue,
  id: id,
  isDisabled: isDisabled,
  label: label,
  name: name,
  transform: Number
}), ({
  fieldProps: {
    isInvalid,
    isRequired,
    ...fieldProps
  },
  ...rest
}) =>
// isInvalid and isRequired are specifically invalid for range inputs
children({
  fieldProps,
  ...rest
}));
export default RangeField;
