// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/no-html-heading/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createRule = require("../utils/create-rule");
var _nodeTypes = require("./node-types");
var rule = (0, _createRule.createLintRule)({
  meta: {
    name: 'no-html-heading',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML heading elements in favor of Atlassian Design System heading components.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlHeading: "This <{{ name }}> should be replaced with a heading component from the Atlassian Design System. ADS headings include ensure accessible implementations and provide access to ADS styling features like design tokens."
    }
  },
  create: function create(context) {
    return {
      // transforms styled.<heading>(...) usages
      CallExpression: function CallExpression(node) {
        _nodeTypes.StyledComponent.lint(node, {
          context: context
        });
      },
      // transforms <heading css={...}> usages
      JSXElement: function JSXElement(node) {
        _nodeTypes.JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
var _default = exports.default = rule;
