// @atlaskit/eslint-plugin-design-system/dist/es2019/rules/no-html-image/index.js
import { createLintRule } from '../utils/create-rule';
import { JSXElement, StyledComponent } from './node-types';
const rule = createLintRule({
  meta: {
    name: 'no-html-image',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML image elements in favor of the Atlassian Design System image component.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlImage: `This <{{ name }}> should be replaced with the image component from the Atlassian Design System. ADS images ensure accessible implementations, and provide access to ADS styling features like design tokens.`
    }
  },
  create(context) {
    return {
      // transforms styled.<anchor>(...) usages
      CallExpression(node) {
        StyledComponent.lint(node, {
          context
        });
      },
      // transforms <anchor css={...}> usages
      JSXElement(node) {
        JSXElement.lint(node, {
          context
        });
      }
    };
  }
});
export default rule;
