// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/use-latest-xcss-syntax-typography/linters/restricted-capitalisation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RestrictedCapitalisation = void 0;
var _eslintCodemodUtils = require("eslint-codemod-utils");
var _common = require("./common");
/* eslint-disable @repo/internal/react/require-jsdoc */

var RestrictedCapitalisation = exports.RestrictedCapitalisation = {
  lint: function lint(node, _ref) {
    var context = _ref.context,
      config = _ref.config;
    if (RestrictedCapitalisation._check(node, {
      context: context,
      config: config
    })) {
      context.report({
        node: node,
        messageId: 'noRestrictedCapitalisation'
      });
    }
  },
  _check: function _check(node, _ref2) {
    var config = _ref2.config;
    if (!config.patterns.includes('restricted-capitalisation')) {
      return false;
    }

    // Prevent text transform being used to uppercase all characters
    if ((0, _common.isPropertyName)(node, 'textTransform') && (0, _eslintCodemodUtils.isNodeOfType)(node.parent, 'Property') && (0, _eslintCodemodUtils.isNodeOfType)(node.parent.value, 'Literal')) {
      return node.parent.value.value === 'uppercase';
    }
    return true;
  }
};
