// @atlaskit/eslint-plugin-design-system/dist/esm/rules/use-tokens-typography/transformers/banned-properties.js
/* eslint-disable @repo/internal/react/require-jsdoc */

import { isNodeOfType } from 'eslint-codemod-utils';
import { isDecendantOfStyleBlock, isDecendantOfType } from '../../utils/is-node';
export var BannedProperties = {
  lint: function lint(node, _ref) {
    var context = _ref.context,
      config = _ref.config;
    // Check whether all criteria needed to make a transformation are met
    var success = BannedProperties._check(node, {
      context: context,
      config: config
    });
    if (success) {
      return context.report({
        node: node,
        messageId: 'noBannedProperties',
        data: {
          property: isNodeOfType(node.key, 'Identifier') ? node.key.name : 'this property'
        }
      });
    }
  },
  _check: function _check(node, _ref2) {
    var config = _ref2.config;
    if (!config.patterns.includes('banned-properties')) {
      return false;
    }
    if (!isNodeOfType(node, 'Property')) {
      return false;
    }
    if (!isDecendantOfStyleBlock(node) && !isDecendantOfType(node, 'JSXExpressionContainer')) {
      return false;
    }
    return true;
  }
};
