// @atlaskit/eslint-plugin-design-system/dist/esm/rules/use-tokens-typography/transformers/font-weight.js
/* eslint-disable @repo/internal/react/require-jsdoc */

import { isNodeOfType } from 'eslint-codemod-utils';
import { getSourceCode } from '@atlaskit/eslint-utils/context-compat';
import { Root } from '../../../ast-nodes';
import { getNodeSource } from '../../utils/get-node-source';
import { isDecendantOfStyleBlock, isDecendantOfType } from '../../utils/is-node';
import { findFontWeightTokenForValue, insertTokensImport } from '../utils';
export var FontWeight = {
  lint: function lint(node, _ref) {
    var context = _ref.context,
      config = _ref.config;
    // Check whether all criteria needed to make a transformation are met
    var success = FontWeight._check(node, {
      context: context,
      config: config
    });
    if (success) {
      return context.report({
        node: node,
        messageId: 'noRawFontWeightValues',
        fix: FontWeight._fix(node, context)
      });
    }
  },
  _check: function _check(node, _ref2) {
    var context = _ref2.context,
      config = _ref2.config;
    if (!config.patterns.includes('font-weight')) {
      return false;
    }
    if (!isNodeOfType(node, 'Property')) {
      return false;
    }
    if (!isDecendantOfStyleBlock(node) && !isDecendantOfType(node, 'JSXExpressionContainer')) {
      return false;
    }
    var isFontWeightProperty = isNodeOfType(node.key, 'Identifier') && node.key.name === 'fontWeight';
    var valueNodeSource = getNodeSource(getSourceCode(context), node.value);
    if (isFontWeightProperty && valueNodeSource.match(/(font\.weight.|inherit)/)) {
      return false;
    }
    return true;
  },
  _fix: function _fix(node, context) {
    return function (fixer) {
      var _findFontWeightTokenF;
      var fixes = [];

      // Type assertions to force the correct node type
      if (!isNodeOfType(node.value, 'Literal')) {
        return fixes;
      }
      if (!node.value.raw) {
        return fixes;
      }

      // Replace raw value with token if there is a token match
      var matchingToken = (_findFontWeightTokenF = findFontWeightTokenForValue(String(node.value.value))) === null || _findFontWeightTokenF === void 0 ? void 0 : _findFontWeightTokenF.tokenName;
      if (!matchingToken) {
        return fixes;
      }
      var fontWeightValueFix = fixer.replaceText(node.value, "token('".concat(matchingToken, "')"));
      fixes.push(fontWeightValueFix);

      // Add import if it doesn't exist
      var body = getSourceCode(context).ast.body;
      var tokensImportDeclarations = Root.findImportsByModule(body, '@atlaskit/tokens');

      // If there is more than one `@atlaskit/tokens` import, then it becomes difficult to determine which import to transform
      if (tokensImportDeclarations.length > 1) {
        return fixes;
      }
      var tokensImportDeclaration = tokensImportDeclarations[0];
      if (!tokensImportDeclaration) {
        fixes.push(insertTokensImport(body, fixer));
      }
      return fixes;
    };
  }
};
