// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/no-html-select/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createRule = require("../utils/create-rule");
var _nodeTypes = require("./node-types");
var rule = (0, _createRule.createLintRule)({
  meta: {
    name: 'no-html-select',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML select elements in favor of the Atlassian Design System select component.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlSelect: "This <{{ name }}> should be replaced with the select component from the Atlassian Design System. ADS select components have event tracking, ensure accessible implementations, and provide access to ADS styling features like design tokens."
    }
  },
  create: function create(context) {
    return {
      // transforms styled.<anchor>(...) usages
      CallExpression: function CallExpression(node) {
        _nodeTypes.StyledComponent.lint(node, {
          context: context
        });
      },
      // transforms <anchor css={...}> usages
      JSXElement: function JSXElement(node) {
        _nodeTypes.JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
var _default = exports.default = rule;
