// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/no-html-button/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createRule = require("../utils/create-rule");
var _nodeTypes = require("./node-types");
var rule = (0, _createRule.createLintRule)({
  meta: {
    name: 'no-html-button',
    type: 'suggestion',
    docs: {
      description: 'Discourage direct usage of HTML button elements in favor of Atlassian Design System button components.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlButton: "This {{ name }} should be replaced with a button component from the Atlassian Design System, such as the \"Button\" component when suitable. For custom buttons use the \"Pressable\" primitive. ADS buttons include event tracking, ensure accessible implementations, and provide access to ADS styling features like design tokens."
    }
  },
  create: function create(context) {
    return {
      // transforms styled.<button>(...) usages
      CallExpression: function CallExpression(node) {
        _nodeTypes.StyledComponent.lint(node, {
          context: context
        });
      },
      // transforms <button css={...}> usages
      JSXElement: function JSXElement(node) {
        _nodeTypes.JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
var _default = exports.default = rule;
