// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/use-primitives/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "convertASTObjectExpressionToJSObject", {
  enumerable: true,
  get: function get() {
    return _convertAstObjectExpressionToJsObject.convertASTObjectExpressionToJSObject;
  }
});
Object.defineProperty(exports, "getAttributeValueIdentifier", {
  enumerable: true,
  get: function get() {
    return _getAttributeValueIdentifier.getAttributeValueIdentifier;
  }
});
Object.defineProperty(exports, "getFunctionArgumentAtPos", {
  enumerable: true,
  get: function get() {
    return _getFunctionArgumentAtPos.getFunctionArgumentAtPos;
  }
});
Object.defineProperty(exports, "getJSXAttributeByName", {
  enumerable: true,
  get: function get() {
    return _getJsxAttributeByName.getJSXAttributeByName;
  }
});
Object.defineProperty(exports, "getVariableDefinitionValue", {
  enumerable: true,
  get: function get() {
    return _getVariableDefinitionValue.getVariableDefinitionValue;
  }
});
Object.defineProperty(exports, "getVariableUsagesCount", {
  enumerable: true,
  get: function get() {
    return _getVariableUsageCount.getVariableUsagesCount;
  }
});
Object.defineProperty(exports, "isFunctionNamed", {
  enumerable: true,
  get: function get() {
    return _isFunctionNamed.isFunctionNamed;
  }
});
Object.defineProperty(exports, "isValidCssPropertiesToTransform", {
  enumerable: true,
  get: function get() {
    return _isValidCssPropertiesToTransform.isValidCssPropertiesToTransform;
  }
});
Object.defineProperty(exports, "updateJSXAttributeByName", {
  enumerable: true,
  get: function get() {
    return _updateJsxAttributeByName.updateJSXAttributeByName;
  }
});
var _convertAstObjectExpressionToJsObject = require("./convert-ast-object-expression-to-js-object");
var _getAttributeValueIdentifier = require("./get-attribute-value-identifier");
var _getFunctionArgumentAtPos = require("./get-function-argument-at-pos");
var _getJsxAttributeByName = require("./get-jsx-attribute-by-name");
var _getVariableDefinitionValue = require("./get-variable-definition-value");
var _getVariableUsageCount = require("./get-variable-usage-count");
var _isFunctionNamed = require("./is-function-named");
var _isValidCssPropertiesToTransform = require("./is-valid-css-properties-to-transform");
var _updateJsxAttributeByName = require("./update-jsx-attribute-by-name");
