// @atlaskit/eslint-plugin-design-system/dist/esm/rules/no-emotion-primitives/index.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import { literal } from 'eslint-codemod-utils';
import { createLintRule } from '../utils/create-rule';
var schema = {
  type: 'array',
  items: [{
    type: 'object',
    properties: {
      autofix: {
        type: 'boolean'
      }
    },
    additionalProperties: false
  }]
};
var defaultConfig = {
  autofix: false
};
function readConfig(context) {
  var _ref = context.options,
    _ref2 = _slicedToArray(_ref, 1),
    config = _ref2[0];
  return Object.assign({}, defaultConfig, config);
}
export var rule = createLintRule({
  meta: {
    name: 'no-emotion-primitives',
    type: 'problem',
    fixable: 'code',
    docs: {
      description: 'Ensures usage of Compiled Primitives import instead of Emotion entrypoint.',
      severity: 'warn',
      recommended: false
    },
    messages: {
      'no-emotion-primitives': 'Use @atlaskit/primitives/compiled instead of @atlaskit/primitives. Refer to go/akcss for more information.'
    },
    schema: schema
  },
  create: function create(context) {
    var config = readConfig(context);
    return {
      ImportDeclaration: function ImportDeclaration(node) {
        var importSource = node.source.value;
        if (importSource !== '@atlaskit/primitives') {
          return;
        }
        context.report({
          node: node.source,
          messageId: 'no-emotion-primitives',
          fix: function fix(fixer) {
            if (!config.autofix) {
              return null;
            }
            var newSource = literal('@atlaskit/primitives/compiled');
            return fixer.replaceText(node.source, newSource.raw || "'@atlaskit/primitives/compiled'");
          }
        });
      }
    };
  }
});
export default rule;
