// @atlaskit/eslint-plugin-design-system/dist/esm/rules/use-primitives/transformers/compiled-styled/index.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
/* eslint-disable @repo/internal/react/require-jsdoc */

import { isNodeOfType } from 'eslint-codemod-utils';
import { getScope, getSourceCode } from '@atlaskit/eslint-utils/context-compat';
import * as ast from '../../../../ast-nodes';
import { isValidCssPropertiesToTransform } from '../../utils';
import { validateStyles } from '../../utils/validate-styles';
import { convertJsxCallSite } from './convert-jsx-call-site';
import { convertStyledComponentToXcss } from './convert-styled-component-call-to-jsx';
import { findValidJsxUsageToTransform } from './find-valid-jsx-usage-to-transform';
import { findValidStyledComponentCall } from './find-valid-styled-component-call';
export var CompiledStyled = {
  lint: function lint(node, _ref) {
    var context = _ref.context,
      config = _ref.config;
    if (!isNodeOfType(node, 'CallExpression')) {
      return;
    }

    // Check whether all criteria needed to make a transformation are met
    var _CompiledStyled$_chec = CompiledStyled._check(node, {
        context: context,
        config: config
      }),
      success = _CompiledStyled$_chec.success,
      refs = _CompiledStyled$_chec.refs;
    if (!success || !refs) {
      return;
    }
    context.report({
      node: refs.styles,
      messageId: 'preferPrimitivesBox',
      suggest: [{
        desc: "Convert ".concat(ast.JSXElement.getName(refs.jsxElement), " to Box"),
        fix: CompiledStyled._fix(refs, context)
      }]
    });
  },
  _check: function _check(node, _ref2) {
    var context = _ref2.context,
      config = _ref2.config;
    if (!config.patterns.includes('compiled-styled-object')) {
      return {
        success: false
      };
    }
    if (!isNodeOfType(node, 'CallExpression')) {
      return {
        success: false
      };
    }
    var styledComponentVariableRef = findValidStyledComponentCall(node);
    if (!styledComponentVariableRef || !isNodeOfType(styledComponentVariableRef.id, 'Identifier') || !(config.patterns.includes('string-style-property-fix') ? validateStyles(node, config) : isValidCssPropertiesToTransform(node, config))) {
      return {
        success: false
      };
    }
    var anyOrder = config.patterns.includes('jsx-order-fix');
    var styledComponentJsxRef = findValidJsxUsageToTransform(styledComponentVariableRef.id.name, getScope(context, node), anyOrder);
    if (!styledComponentJsxRef) {
      return {
        success: false
      };
    }
    if (!isNodeOfType(styledComponentJsxRef.parent, 'JSXElement')) {
      return {
        success: false
      };
    }
    var importDeclaration = ast.Root.findImportsByModule(getSourceCode(context).ast.body, '@atlaskit/primitives');

    // If there is more than one `@atlaskit/primitives` import, then it becomes difficult to determine which import to transform
    if (importDeclaration.length > 1) {
      return {
        success: false
      };
    }
    return {
      success: true,
      refs: {
        styles: styledComponentVariableRef,
        jsxElement: styledComponentJsxRef.parent
      }
    };
  },
  /**
   * All required validation steps have been taken care of before this
   * transformer is called, so it just goes ahead providing all necessary fixes
   */
  _fix: function _fix(refs, context) {
    return function (fixer) {
      // generates the new variable name: MyComponent -> myComponentStyles
      var calculatedStylesVariableName = isNodeOfType(refs.styles.id, 'Identifier') && "".concat(refs.styles.id.name.replace(refs.styles.id.name[0], refs.styles.id.name[0].toLowerCase()), "Styles");
      if (!calculatedStylesVariableName) {
        return [];
      }
      var importFixes = ast.Root.upsertNamedImportDeclaration({
        module: '@atlaskit/primitives',
        specifiers: ['Box', 'xcss']
      }, context, fixer);
      var stylesFixes = convertStyledComponentToXcss(refs.styles, calculatedStylesVariableName, fixer);
      var jsxFixes = convertJsxCallSite(refs.jsxElement, calculatedStylesVariableName, fixer);
      return [importFixes].concat(_toConsumableArray(stylesFixes), _toConsumableArray(jsxFixes)).filter(function (fix) {
        return Boolean(fix);
      }); // Some of the transformers can return arrays with undefined, so filter them out
    };
  }
};
