// @atlaskit/eslint-plugin-design-system/dist/esm/rules/no-html-heading/index.js
import { createLintRule } from '../utils/create-rule';
import { JSXElement as _JSXElement, StyledComponent } from './node-types';
var rule = createLintRule({
  meta: {
    name: 'no-html-heading',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML heading elements in favor of Atlassian Design System heading components.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlHeading: "This <{{ name }}> should be replaced with a heading component from the Atlassian Design System. ADS headings include ensure accessible implementations and provide access to ADS styling features like design tokens."
    }
  },
  create: function create(context) {
    return {
      // transforms styled.<heading>(...) usages
      CallExpression: function CallExpression(node) {
        StyledComponent.lint(node, {
          context: context
        });
      },
      // transforms <heading css={...}> usages
      JSXElement: function JSXElement(node) {
        _JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
export default rule;
