// @atlaskit/eslint-plugin-design-system/dist/esm/index.codegen.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * THIS FILE WAS CREATED VIA CODEGEN DO NOT MODIFY {@see http://go/af-codegen}
 * @codegen <<SignedSource::1ca7a384145f3bf3577ede9b88907a52>>
 * @codegenCommand yarn workspace @atlaskit/eslint-plugin-design-system codegen
 */

import allFlat from './presets/all-flat.codegen';
import all from './presets/all.codegen';
import recommendedFlat from './presets/recommended-flat.codegen';
import recommended from './presets/recommended.codegen';
import { rules } from './rules/index.codegen';

// this uses require because not all node versions this package supports use the same import assertions/attributes
// eslint-disable-next-line import/no-extraneous-dependencies
var pkgJson = require('@atlaskit/eslint-plugin-design-system/package.json');
var version = pkgJson.version,
  name = pkgJson.name;
var meta = {
  name: name,
  version: version
};
export var plugin = {
  meta: meta,
  rules: rules,
  configs: {
    all: all,
    'all/flat': _objectSpread(_objectSpread({}, allFlat), {}, {
      plugins: _objectSpread(_objectSpread({}, allFlat.plugins), {}, {
        get '@atlaskit/design-system'() {
          return plugin;
        }
      })
    }),
    recommended: recommended,
    'recommended/flat': _objectSpread(_objectSpread({}, recommendedFlat), {}, {
      plugins: _objectSpread(_objectSpread({}, recommendedFlat.plugins), {}, {
        get '@atlaskit/design-system'() {
          return plugin;
        }
      })
    })
  }
};
var configs = plugin.configs;
export { configs, meta, rules };
export default plugin;
