// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/utils/is-node.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isPropertyKey = exports.isDecendantOfXcssBlock = exports.isDecendantOfType = exports.isDecendantOfSvgElement = exports.isDecendantOfStyleJsxAttribute = exports.isDecendantOfStyleBlock = exports.isDecendantOfPrimitive = exports.isDecendantOfGlobalToken = exports.isCssInJsTemplateNode = exports.isCssInJsObjectNode = exports.isCssInJsCallNode = exports.isChildOfType = void 0;
var _eslintCodemodUtils = require("eslint-codemod-utils");
var _contextCompat = require("@atlaskit/eslint-utils/context-compat");
var _isSupportedImport = require("@atlaskit/eslint-utils/is-supported-import");
var _astNodes = require("../../ast-nodes");
var _isDecendantOfGlobalToken = exports.isDecendantOfGlobalToken = function isDecendantOfGlobalToken(node) {
  if ((0, _eslintCodemodUtils.isNodeOfType)(node, 'CallExpression') && (0, _eslintCodemodUtils.isNodeOfType)(node.callee, 'Identifier') && (node.callee.name === 'token' || node.callee.name === 'getTokenValue')) {
    return true;
  }
  if (node.parent) {
    return _isDecendantOfGlobalToken(node.parent);
  }
  return false;
};
var _isDecendantOfType = exports.isDecendantOfType = function isDecendantOfType(node, type) {
  var skipNode = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  if (!skipNode && node.type === type) {
    return true;
  }
  if (node.parent) {
    return _isDecendantOfType(node.parent, type, false);
  }
  return false;
};
var isPropertyKey = exports.isPropertyKey = function isPropertyKey(node) {
  if ((0, _eslintCodemodUtils.isNodeOfType)(node, 'Identifier') && _isDecendantOfType(node, 'Property')) {
    var parent = node.parent;
    return node === parent.key || parent.shorthand;
  }
  return false;
};
var _isDecendantOfStyleJsxAttribute = exports.isDecendantOfStyleJsxAttribute = function isDecendantOfStyleJsxAttribute(node) {
  if ((0, _eslintCodemodUtils.isNodeOfType)(node, 'JSXAttribute')) {
    return true;
  }
  if (node.parent) {
    return _isDecendantOfStyleJsxAttribute(node.parent);
  }
  return false;
};
var _isDecendantOfSvgElement = exports.isDecendantOfSvgElement = function isDecendantOfSvgElement(node) {
  if ((0, _eslintCodemodUtils.isNodeOfType)(node, 'JSXElement')) {
    // @ts-ignore
    if (node.openingElement.name.name === 'svg') {
      return true;
    }
  }
  if (node.parent) {
    return _isDecendantOfSvgElement(node.parent);
  }
  return false;
};
var _isDecendantOfPrimitive = exports.isDecendantOfPrimitive = function isDecendantOfPrimitive(node, context) {
  var primitivesToCheck = ['Box', 'Text'];
  if ((0, _eslintCodemodUtils.isNodeOfType)(node, 'JSXElement')) {
    // @ts-ignore
    if (primitivesToCheck.includes(node.openingElement.name.name)) {
      var importDeclaration = _astNodes.Root.findImportsByModule((0, _contextCompat.getSourceCode)(context).ast.body, ['@atlaskit/primitives', '@atlaskit/primitives/box', '@atlaskit/primitives/text', '@atlaskit/primitives/compiled']);
      if (importDeclaration.length) {
        return true;
      }
    }
  }
  if (node.parent) {
    return _isDecendantOfPrimitive(node.parent, context);
  }
  return false;
};
var cssInJsCallees = ['css', 'styled', 'styled2'];
var isCssInJsTemplateNode = exports.isCssInJsTemplateNode = function isCssInJsTemplateNode(node) {
  return (node === null || node === void 0 ? void 0 : node.type) === 'TaggedTemplateExpression' && node.tag.type === 'MemberExpression' && node.tag.object.type === 'Identifier' && node.tag.object.name === 'styled';
};
var isCssInJsCallNode = exports.isCssInJsCallNode = function isCssInJsCallNode(node) {
  return (node === null || node === void 0 ? void 0 : node.type) === 'CallExpression' && node.callee.type === 'Identifier' && cssInJsCallees.includes(node.callee.name);
};
var isCssInJsObjectNode = exports.isCssInJsObjectNode = function isCssInJsObjectNode(node) {
  return (node === null || node === void 0 ? void 0 : node.type) === 'CallExpression' && node.callee.type === 'MemberExpression' && node.callee.object.type === 'Identifier' && cssInJsCallees.includes(node.callee.object.name);
};
var _isDecendantOfXcssBlock = exports.isDecendantOfXcssBlock = function isDecendantOfXcssBlock(node, referencesInScope, importSources) {
  // xcss contains types for all properties that accept tokens, so ignore xcss for linting as it will report false positives
  if (node.type === 'CallExpression' && (0, _isSupportedImport.isXcss)(node.callee, referencesInScope, importSources)) {
    return true;
  }
  if (node.parent) {
    return _isDecendantOfXcssBlock(node.parent, referencesInScope, importSources);
  }
  return false;
};
var _isDecendantOfStyleBlock = exports.isDecendantOfStyleBlock = function isDecendantOfStyleBlock(node) {
  if (node.type === 'VariableDeclarator') {
    if (node.id.type !== 'Identifier') {
      return false;
    }
    if (
    // @ts-ignore typeAnnotation is not defined by types
    node.id.typeAnnotation &&
    // @ts-ignore typeAnnotation is not defined by types
    node.id.typeAnnotation.typeAnnotation.type === 'GenericTypeAnnotation' &&
    // @ts-ignore typeAnnotation is not defined by types
    node.id.typeAnnotation.typeAnnotation.id.type === 'Identifier') {
      // @ts-ignore typeAnnotation is not defined by types
      var typeName = node.id.typeAnnotation.typeAnnotation.id.name;
      var hasCSSType = ['CSSProperties', 'CSSObject'].some(function (el) {
        return typeName.includes(el);
      });
      if (hasCSSType) {
        return true;
      }
    }

    // @ts-ignore Name is not defined in types
    var varName = node.id.name.toLowerCase();
    return ['style', 'css', 'theme'].some(function (el) {
      return varName.includes(el);
    });
  }
  if (isCssInJsCallNode(node) || isCssInJsObjectNode(node) || isCssInJsTemplateNode(node)) {
    return true;
  }
  if (node.type === 'TaggedTemplateExpression' && node.tag.type === 'Identifier' && node.tag.name === 'css') {
    return true;
  }
  if (node.parent) {
    return _isDecendantOfStyleBlock(node.parent);
  }
  return false;
};
var isChildOfType = exports.isChildOfType = function isChildOfType(node, type) {
  return (0, _eslintCodemodUtils.isNodeOfType)(node.parent, type);
};
