// @atlaskit/eslint-utils/dist/es2019/allowed-function-calls/index.js
// allowedFunctionCalls is an ESLint rule option used by `no-invalid-css-map` and `no-unsafe-values` to whitelist certain function calls from restrictions imposed by those rules.
//
// Expected usage:
//     {
//         // other options for your ESLint rule...
//         // ...
//         allowedFunctionCalls: [
//             ['@atlaskit/tokens', 'token'],
//             ['package-name-here', 'function-name-here'],
//         ],
//     }
//
// You can use these utility functions to process the allowedFunctionsCall option for use in your own rule.
//
// Note that whitelisting default exports is not supported (yet?).

import merge from 'lodash/merge';
import { defaultAllowedDynamicKeys, defaultAllowedValues } from './default-allowed';
export function isAllowListedVariable({
  allowList,
  variable
}) {
  const definitions = variable.defs;
  if (!definitions) {
    return false;
  }
  return definitions.every(definition => {
    var _allowList$packageNam;
    // We add some restrictions to keep this simple...
    // Forbid non-imported functions
    if (definition.type !== 'ImportBinding') {
      return false;
    }

    // Forbid default imports (e.g. `import React from 'react'`)
    if (definition.node.type !== 'ImportSpecifier') {
      return false;
    }
    const packageName = definition.parent.source.value;
    const importName = definition.node.imported.name;
    return typeof packageName === 'string' && ((_allowList$packageNam = allowList[packageName]) === null || _allowList$packageNam === void 0 ? void 0 : _allowList$packageNam.includes(importName));
  });
}

/**
 * Differs to `Object.fromEntries` in how it handles multiple entries with the same key.
 */
function collectEntries(entries) {
  return entries.reduce((allowList, [packageName, importName]) => {
    var _allowList$packageNam2;
    const currentValue = (_allowList$packageNam2 = allowList[packageName]) !== null && _allowList$packageNam2 !== void 0 ? _allowList$packageNam2 : [];
    return {
      ...allowList,
      [packageName]: [...currentValue, importName]
    };
  }, {});
}
function getAllowList({
  options,
  allowListKey,
  defaultAllowList
}) {
  if (options.length === 0 || options[0][allowListKey] === undefined) {
    return defaultAllowList;
  }

  // We are not checking the options type.
  // We assume that ESLint is enforcing the rule's `schema` correctly.
  // Intentionally passing `{}` as first arg as this function mutates
  return merge({}, defaultAllowList, collectEntries(options[0][allowListKey]));
}
export const getAllowedFunctionCalls = options => getAllowList({
  options,
  allowListKey: 'allowedFunctionCalls',
  defaultAllowList: defaultAllowedValues
});
export const getAllowedDynamicKeys = options => getAllowList({
  options,
  allowListKey: 'allowedDynamicKeys',
  defaultAllowList: defaultAllowedDynamicKeys
});
