// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/no-html-image/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _createRule = require("../utils/create-rule");
var _nodeTypes = require("./node-types");
var rule = (0, _createRule.createLintRule)({
  meta: {
    name: 'no-html-image',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML image elements in favor of the Atlassian Design System image component.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlImage: "This <{{ name }}> should be replaced with the image component from the Atlassian Design System. ADS images ensure accessible implementations, and provide access to ADS styling features like design tokens."
    }
  },
  create: function create(context) {
    return {
      // transforms styled.<anchor>(...) usages
      CallExpression: function CallExpression(node) {
        _nodeTypes.StyledComponent.lint(node, {
          context: context
        });
      },
      // transforms <anchor css={...}> usages
      JSXElement: function JSXElement(node) {
        _nodeTypes.JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
var _default = exports.default = rule;
