// @atlaskit/eslint-plugin-design-system/dist/esm/rules/use-primitives/index.js
import { createLintRule } from '../utils/create-rule';
import { getConfig, VALID_PATTERNS } from './config';
import { CompiledStyled, EmotionCSS } from './transformers';
var boxDocsUrl = 'https://atlassian.design/components/primitives/box';
var rule = createLintRule({
  meta: {
    name: 'use-primitives',
    type: 'suggestion',
    fixable: 'code',
    hasSuggestions: true,
    docs: {
      description: 'Encourage the usage of primitives components.',
      recommended: false,
      severity: 'warn'
    },
    schema: [{
      type: 'object',
      properties: {
        patterns: {
          type: 'array',
          maxLength: VALID_PATTERNS.length,
          items: {
            type: 'string',
            enum: VALID_PATTERNS
          }
        }
      },
      additionalProperties: false
    }],
    messages: {
      preferPrimitivesBox: "This element can be replaced with a \"Box\" primitive. See ".concat(boxDocsUrl, " for additional guidance.")
    }
  },
  create: function create(context) {
    // TODO: JFP-2823 - this type cast was added due to Jira's ESLint v9 migration
    var config = getConfig(context.options[0]);
    return {
      // transforms styled.<html>(...) usages
      CallExpression: function CallExpression(node) {
        CompiledStyled.lint(node, {
          context: context,
          config: config
        });
      },
      // transforms <div css={...}> usages
      JSXElement: function JSXElement(node) {
        EmotionCSS.lint(node, {
          context: context,
          config: config
        });
      }
    };
  }
});
export default rule;
