// @atlaskit/eslint-plugin-design-system/dist/cjs/rules/no-keyframes-tagged-template-expression/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _isSupportedImport = require("@atlaskit/eslint-utils/is-supported-import");
var _createNoTaggedTemplateExpressionRule = require("../utils/create-no-tagged-template-expression-rule");
var _createRule = require("../utils/create-rule");
var rule = (0, _createRule.createLintRule)({
  meta: {
    name: 'no-keyframes-tagged-template-expression',
    fixable: 'code',
    type: 'problem',
    docs: {
      description: 'Disallows any `keyframe` tagged template expressions that originate from Emotion, Styled Components or Compiled',
      recommended: false,
      severity: 'error'
    },
    messages: {
      unexpectedTaggedTemplate: 'Unexpected `keyframes` tagged template expression'
    },
    schema: _createNoTaggedTemplateExpressionRule.noTaggedTemplateExpressionRuleSchema
  },
  create: (0, _createNoTaggedTemplateExpressionRule.createNoTaggedTemplateExpressionRule)(_isSupportedImport.isKeyframes, 'unexpectedTaggedTemplate')
});
var _default = exports.default = rule;
