// @atlaskit/eslint-plugin-design-system/dist/esm/rules/no-html-code/index.js
import { createLintRule } from '../utils/create-rule';
import { JSXElement as _JSXElement, StyledComponent } from './node-types';
var rule = createLintRule({
  meta: {
    name: 'no-html-code',
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      description: 'Discourage direct usage of HTML code elements in favor of the Atlassian Design System code component.',
      recommended: true,
      severity: 'warn'
    },
    messages: {
      noHtmlCode: "This <{{ name }}> should be replaced with the code component from the Atlassian Design System. The ADS code component ensures accessible implementations and consistent ADS styling."
    }
  },
  create: function create(context) {
    return {
      // transforms styled.<anchor>(...) usages
      CallExpression: function CallExpression(node) {
        StyledComponent.lint(node, {
          context: context
        });
      },
      // transforms <anchor css={...}> usages
      JSXElement: function JSXElement(node) {
        _JSXElement.lint(node, {
          context: context
        });
      }
    };
  }
});
export default rule;
